blockchain: add tests for BlockStore

odeke-em opened the pull request:
Add tests to test store, to the fullest reasonable extent for
paths that can taken by input arguments altering internal behavior.

melekes commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,263 @@
+			_ = blockchain.NewBlockStore(db)
+			return nil, nil
+		})
+		assert.NotNil(t, err, "#%d panicCauser: %q expected a panic", i, panicCauser)
```
there is an https://godoc.org/github.com/stretchr/testify/assert#Panics

odeke-em replied:
Nice, thank you for the reference, I hadn't seen that function. I'll skip using it though for this test particularly because for some of these we want panics, for others we don't and constructing the code with assert.Panics might be a little hairy, I haven't seen an easy way of doing it :)

ebuchman replied:
arent all three of these test cases expecting a panic ?

odeke-em replied:
@ebuchman in this isolated test with these 3 test cases, yes all those 3 expect panics, but doFn is also used throughout the file in other cases that don't necessarily want to panic.

melekes commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,263 @@
+			bBlock := bs.LoadBlock(tuple.block.Height)
+			bBlockMeta := bs.LoadBlockMeta(tuple.block.Height)
+			bSeenCommit := bs.LoadSeenCommit(tuple.block.Height)
+			if tuple.eraseDBForCommit {
```
replayed in the associated PR. I think we should not erase db by calling `Close()`. feels wrong

adrianbrink replied:
Can you reattach to the in-mem db after closing it?

odeke-em replied:
@melekes thank you for the advocacy, I've reflected on your comments in the PR https://github.com/tendermint/tmlibs/pull/56 and I concede with your thoughts :)

@adrianbrink yes we can :) However, I'll instead invoke an iterator and delete all the content in the DB for this test instead of invoking .Close()

odeke-em replied:
I've updated the code, no longer need to invoke db.Close(), I now have a helper that literally erases the DB, PTAL @melekes and @adrianbrink, thanks.

melekes commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,268 @@
+		wantErr          bool
+		wantPanic        bool
+		eraseDBForCommit bool
+		skipReason       string
```
can't see where this is used

odeke-em replied:
Ahh crap, I forgot to remove it after the last commit, I'll update that, thanks.

melekes commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,267 @@
+
+	// Setup, test data
+	part2 := &types.Part{Index: 1, Bytes: data.Bytes([]byte{
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x00,
```
how did you generate these? or it's just random bytes?

odeke-em replied:
So I generated them by
```go
package main

import (
	"bytes"
	"log"

	"github.com/tendermint/go-wire"
	"github.com/tendermint/tendermint/types"
)

func main() {
	buf := new(bytes.Buffer)
	var n int
	var err error

	block := &types.Block{
		Header: &types.Header{
			Height:  1,
			NumTxs:  100,
			ChainID: "block_test",
		},
		LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
	}
	wire.WriteBinary(block, buf, &n, &err)
	log.Printf("% x", buf.Bytes())
}
```
and then split the bytes over two parts.

However, I didn't include the generation code in the test, lest it caused confusion.

melekes replied:
But if something changes, how a next person supposed to change this stuff? Maybe we should at least add a comment?

adrianbrink replied:
Let's include the generation code as well and ideally generate them on each test run.

odeke-em replied:
Roger that, I'll add that comment in in a second. @adrianbrink am not too sure about regenerating it on each test run as that makes the test harder to grok, however a note on how to generate the parts bytes.

odeke-em commented:
Alright, I've updated the test to also include how to generate the parts' data, PTAL.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+func doFn(fn func() (interface{}, error)) (res interface{}, err error, panicErr error) {
```
is this worth it? why not just let it panic in the test?

odeke-em replied:
I believe we need this function. The point of it is to capture the panic and relay it back to whatever caller needs to check if we expected a panic or not -- for example if the blockMeta content is messed up vs when a block is successfully read.
If we just let the test panic, we wouldn't be asserting expected behavior and the tests wouldn't matter. I hope this makes sense.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,269 @@
+
+func eraseDB(db db.DB) {
+	it := db.Iterator()
+	for {
```
can't we just do:

```
for it.Next(){
 db.Delete(it.Key())
}
```

odeke-em replied:
For some reason that was initially buggy on my side and not erasing the last element sometime back, but that now looks functional, thank you for the update.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,269 @@
+	// Setup, test data
+	// The parts' data can be generated by running
+	// the code at https://gist.github.com/odeke-em/9ffac2b5df44595fad7084ece4c9bd98
+	part1 := &types.Part{Index: 0, Bytes: data.Bytes([]byte{
```
i feel like we can simplify this test a bunch. do we really need these part sets? we can derrive part sets directly from blocks, and we have functions to make blocks so we shouldnt need much external madness. 

If we want incompelte/bad part sets, we can mutate them after derriving them from blocks.

odeke-em replied:
@ebuchman there are 3 different partsets that we are testing:
1) validAndCompletePartSet
2) incompletePartSet
3) UncontiguousPartSet

Mutating after deriving IMO makes it even more complex. Could you please help with an example of what you mean?

odeke-em replied:
For reference https://github.com/tendermint/tendermint/pull/677/files/23cb7d7c12ab48d30f23b2590c6bc3d3e72c9e29#diff-2dcc7092e28fd6d98ff9d9ef3b6bdcd1R118
<img width="622" alt="screen shot 2017-10-03 at 9 24 23 pm" src="https://user-images.githubusercontent.com/4898263/31158775-753d8f32-a881-11e7-9db5-60ff38e03448.png">

ebuchman replied:
cant we use block.MakePartSet ?

ebuchman commented:
Can we also take this opportunity to fix up the comments in store.go.

Also might be nice to have some tests about the Height() - ie. we should always be able to fetch block bs.Height(), but bs.Height()+1 should be nil

odeke-em commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,269 @@
+	assert.Equal(t, *bsj, retrBSJ, "expected the retrieved DBs to match")
+}
+
+var blockStoreKey = []byte("blockStore")
```
@ebuchman I was just mocking malicious behavior, but sure, I can move the file to be a part of package `blockchain` instead of `blockchain_test`

odeke-em commented:
@ebuchman by fix, what do you mean? Do you mean add comments? Or review the available comments?

ebuchman commented:
>  Do you mean add comments?

Yeh, just something simple for the exported values

odeke-em commented:
Thanks for the suggestions errbody. I've updated the code and added even more tests for BlockStore plus the requested docs for the exposed functions, PTAL.

ebuchman commented on blockchain/store.go:
```diff
@@ -7,10 +7,10 @@ import (
-	dbm "github.com/tendermint/tmlibs/db"
 	"github.com/tendermint/go-wire"
 	"github.com/tendermint/tendermint/types"
+	. "github.com/tendermint/tmlibs/common"
```
we must beat back these anonymous imports one PR at a time!

zramsay replied:
+++

zramsay replied:
#711

ebuchman commented:
Thanks @odeke-em . I'll make any final edits and merge

ebuchman commented:
@odeke-em thanks for the extensive comments. That said, I feel they are a bit too verbose, with undue emphasis on implementation details. I have updated them in https://github.com/tendermint/tendermint/pull/677/commits/4b393bd0512bbf9819ea1edaf595c36a737a777c - PTAL for style

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,459 @@
+
+	panicCausers := []struct {
+		data    []byte
+		wantErr string
```
this feels way to fragile

ebuchman replied:
can you point to an example of how other large Go projects test their panics ?

odeke-em replied:
Great question. So I picked up this pattern from the Go standard library and here is an exhibit of the usage of `strings.Contains` for substring checks https://gist.github.com/odeke-em/32a71113ee50da57d8a1b861d926d328
Or you can run it in your terminal like this
```shell
$ cd $GOPATH/src && grep -Rn 'strings.Contains' * | grep '.*_test.go' | pbcopy 
```

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,459 @@
+		key  []byte
+		want []byte
+	}{
+		0: {key: []byte("Foo"), want: []byte("Bar")},
```
This doesn't feel like a very useful test. Maybe we can change it to test the Reader in the context of actually reading blocks, rather than simple strings? Ie. the sequence of: 

- marshal block
- set block in db
- GetReader
- ReadBinary
- compare result to original block

odeke-em replied:
@ebuchman good point, however, those sequences/suggestions are covered in the end-to-end tests in other cases. The purpose of this test is to ensure:
a) If we have nil as the value, that we get back a nil Reader
b) A reader if non-nil will not error out if we try to slurp out all the data.

However, if we still can't come to consensus, we can remove this test altogether
since the behavior that it tests is arguably already covered by other tests.

odeke-em replied:
@ebuchman PTAL at my comments above.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,459 @@
+	// Setup, test data
+	// If needed, the parts' data can be generated by running
+	// the code at https://gist.github.com/odeke-em/9ffac2b5df44595fad7084ece4c9bd98
+	part1 = &types.Part{Index: 0, Bytes: data.Bytes([]byte{
```
Feels too fragile. We should use MakeBlock and break it to parts instead of constructing from scratch like this.

odeke-em replied:
I definitely hadn't thought of that, that's way simpler and cleaner. Thank you.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,459 @@
+	}{
+		{
+			block: &types.Block{
+				Header: &types.Header{
```
these headers are all the same, save maybe an odd change. lets make a var for each distinct one and use those.

```
header1 := types.Header{
		Height:  1,
		NumTxs:  100,
		ChainID: "block_test",
}
header2 := header1
header2.Height = 4
```
...

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,459 @@
+	seenCommit1 = &types.Commit{Precommits: []*types.Vote{{Height: 10}}}
+)
+
+func TestBlockStoreSaveLoadBlock(t *testing.T) {
```
Overall this test is super large and unweildly. Think we could break it up into Test..SaveLoadBlock, Test...SaveLoadBlockIncompletePartSet, Test...SaveLoadBlokcUncontiguousPartSet

ebuchman commented:
>  I'll make any final edits and merge

I meant this with respect to the doc comments. Leaving test cleanup to you please :)

ebuchman commented:
Thanks @odeke-em. Your comments all sound good. Indeed, I think mentioning the panic situation once up front is sufficient. Please feel free to make the changes you've suggested and the test changes I suggested when you have time

odeke-em commented:
Alright squad, thanks for the feedback. I've made the tests less fragile, PTAL.

odeke-em commented:
Alright, I've rebased this code to fix a merge conflict that had crept up.

ebuchman commented on blockchain/store_test.go:
```diff
@@ -0,0 +1,424 @@
+		eraseSeenCommitInDB   bool
+	}{
+		{
+			block: &types.Block{
```
there's lots of repetition in laying out these structs that could be compressed

odeke-em replied:
Interesting thought that I hadn't considered. Thanks for raising it, I've updated that.

zramsay commented on blockchain/store.go:
```diff
@@ -23,7 +23,7 @@ There are three types of information stored:
 Currently the precommit signatures are duplicated in the Block parts as
 well as the Commit.  In the future this may change, perhaps by moving
-the Commit data outside the Block.
+the Commit data outside the Block. (TODO)
```
link to an issue?

ebuchman replied:
not priority enough
